java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java

```diff
@@ -68,31 +76,51 @@ public void leaveFile(JavaFileScannerContext context) {
   @Override
   public void visitNode(Tree tree) {
     TryStatementTree tryStatementTree = (TryStatementTree) tree;
+
     withinInterruptingFinally.addFirst(isFinallyInterrupting(tryStatementTree.finallyBlock()));
     for (CatchTree catchTree : tryStatementTree.catches()) {
-      Optional<Type> interruptType = findInterruptingType(catchTree.parameter());
-      if(interruptType.isPresent()) {
-        BlockVisitor blockVisitor = new BlockVisitor(catchTree.parameter().symbol());
-        catchTree.block().accept(blockVisitor);
-        if(!blockVisitor.threadInterrupted && !isWithinInterruptingFinally()) {
-          reportIssue(catchTree.parameter(), "Either re-interrupt this method" +
-            " or rethrow the \""+interruptType.get().name()+"\".");
+      VariableTree catchParameter = catchTree.parameter();
+      List<Type> caughtTypes = getCaughtTypes(catchParameter);
+      Optional<Type> interruptType = caughtTypes.stream().filter(INTERRUPTING_TYPE_PREDICATE).findFirst();
+      if (interruptType.isPresent()) {
+        if (handleInCorrectlyInterruption(catchTree)) {
+          reportIssue(catchParameter, String.format(MESSAGE, interruptType.get().name()));
         }
+        return;
+      } else if (caughtTypes.stream().anyMatch(GENERIC_EXCEPTION_PREDICATE)) {
+        reportIfThrowInterruptInBlock(tryStatementTree.block(), catchTree);
+        return;
       }
     }
   }
 
-  private static Optional<Type> findInterruptingType(VariableTree parameter) {
+  private void reportIfThrowInterruptInBlock(BlockTree blockTree, CatchTree catchTree) {
+    MethodTreeUtils.MethodInvocationCollector collector = new MethodTreeUtils.MethodInvocationCollector(InterruptedExceptionCheck::throwInterruptedException);
+    blockTree.accept(collector);
+    List<Tree> invocationInterrupting = collector.getInvocationTree();
+
+    if (!invocationInterrupting.isEmpty() && handleInCorrectlyInterruption(catchTree)) {
+      reportIssue(catchTree.parameter(), String.format(MESSAGE, "InterruptedException"),
+        invocationInterrupting.stream()
+          .map(t -> new JavaFileScannerContext.Location("Method invocation throwing InterruptedException.", t))
+          .collect(Collectors.toList()),
+        null);
+    }
+  }
+
+  private boolean handleInCorrectlyInterruption(CatchTree catchTree) {
```

Comment:
maybe rename this method to `wasNotInterrupted` or invert condition and name it `wasInterrupted`